src/freebsd.c

```diff
@@ -0,0 +1,71 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <stdint.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+#include <uv.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "../include/fs-ext.h"
+#include "platform.h"
+
+#ifndef RENAME_EXCHANGE
+#define RENAME_EXCHANGE 2
+#endif
+
+int
+fs_ext__try_lock (uv_os_fd_t fd, uint64_t offset, size_t length, fs_ext_lock_type_t type) {
+  if (offset != 0 || length != 0) return UV_EINVAL;
+
+  int res = flock(fd, (type == FS_EXT_WRLOCK ? LOCK_EX : LOCK_SH) | LOCK_NB);
+
+  return res == -1 ? uv_translate_sys_error(errno) : res;
+}
+
+int
+fs_ext__wait_for_lock (uv_os_fd_t fd, uint64_t offset, size_t length, fs_ext_lock_type_t type) {
+  if (offset != 0 || length != 0) return UV_EINVAL;
+
+  int res = flock(fd, type == FS_EXT_WRLOCK ? LOCK_EX : LOCK_SH);
+
+  return res == -1 ? uv_translate_sys_error(errno) : res;
+}
+
+int
+fs_ext__unlock (uv_os_fd_t fd, uint64_t offset, size_t length) {
+  if (offset != 0 || length != 0) return UV_EINVAL;
+
+  int res = flock(fd, LOCK_UN);
+
+  return res == -1 ? uv_translate_sys_error(errno) : res;
+}
+
+int
+fs_ext__trim (uv_os_fd_t fd, uint64_t offset, size_t length) {
+  int res = posix_fallocate(fd, offset, length);
```

Comment:
This will preallocate storage at `offset` for `length`, which is the opposite of what `fs_ext_trim()` does; it _deallocates_ storage at `offset` for `length`, creating a hole in `fd` if possible.

Reply:
We could use https://www.freebsd.org/cgi/man.cgi?query=truncate&sektion=2 but seems it doesnt take an offset. Wdyt?

Reply:
`ftruncate()` will deallocate all storage beyond `length` while we're looking to deallocate storage _in the middle_ of a file. It looks like FreeBSD only fairy recently added support for this: https://www.freebsd.org/status/report-2021-07-2021-09/#_hole_punching_support_on_freebsd

Reply:
The new API is `fspacectl()` (https://reviews.freebsd.org/D28347), available as of FreeBSD 14 (https://docs.freebsd.org/en/books/porters-handbook/versions/).

Reply:
With that in mind, we could do this:

1. If `__FreeBSD_version` is `>= 1400029 ` (introduction of `fspacectl()`), deallocate space at `offset` for `length` using `fspacectl()`.
2. Else, write zeroes at `offset` for `length.

Reply:
Sounds cool!